### ISSUE TITLE
Libraries could not be resolved by the "cdnjs" provider

### ISSUE THREAD

**Issue by duykhiem** [aspnet/LibraryManager#685]
### Describe the bug
We are having an issue with the "cdnjs" provider when building our project in **both Hanoi, Vietnam, and Grimstad, Norway**.
It is happening with **both local build and Azure DevOps Pipeline build**.
**Failed on some random libraries** like jquery@3.4.1, tabulator@4.4.1, tempusdominus-bootstrap-4@5.1.2...

**Error**:

```

Restore operation started
##[error]libman.json(0,0): Error LIB002: The "jquery@3.4.1" library could not be resolved by the "cdnjs" provider
     1>libman.json : error LIB002: The "jquery@3.4.1" library could not be resolved by the "cdnjs" provider 
         One or more libraries failed to restore
         
Build FAILED.

       (LibraryManagerRestore target) -> 
         **libman.json : error LIB002: The "jquery@3.4.1" library could not be resolved by the "cdnjs" provider** 
```

### To Reproduce
1. Using this version of Visual Studio 2022
2. Build the solution
4. See error:
![image](https://user-images.githubusercontent.com/29139700/174572903-9bcba493-21fe-4517-91d7-e63027800370.png)

### Expected behavior
Build and restore client libraries successfully 

### Screenshots
![image](https://user-images.githubusercontent.com/29139700/174573052-b307f452-966d-4f97-ac75-5d1517282fbd.png)
![image](https://user-images.githubusercontent.com/29139700/174573611-dc192c79-c2e4-45e0-acb8-15c591ac2d63.png)
![image](https://user-images.githubusercontent.com/29139700/174573819-d3bb69fe-e8b4-4897-b381-4a1331a27080.png)

### Additional context
- We are using cdnjs as the default provider in [LibMan](https://docs.microsoft.com/en-us/aspnet/core/client-side/libman/libman-vs?view=aspnetcore-6.0)
- It was working perfectly before but **suddenly failed today without any changes related to Libman or package references**

LibMan CLI version: 2.1.161+abc97ecc7d.RR
Visual Studio 2022
Microsoft.Web.LibraryManager.Build" version="2.1.161" targetFramework="net472"

**Comment by duykhiem** [comment]
The workaround is to clear Libman cache
https://github.com/cdnjs/cdnjs/issues/14141#issuecomment-1160220546

**Comment by stebueh** [comment]
possibly related to cdnjs API change, see https://github.com/cdnjs/cdnjs/issues/14140; it's being rolled out partially, which would explain the intermittent failure

**Comment by duykhiem** [comment]
> The workaround is to clear Libman cache [cdnjs/cdnjs#14141 (comment)](https://github.com/cdnjs/cdnjs/issues/14141#issuecomment-1160220546)

Ah. Cleaning cache is just a lucky shot

I think the main issue is related to 
https://github.com/cdnjs/cdnjs/issues/14140

Failed restore/install fall into their 10% production request
https://github.com/cdnjs/cdnjs/issues/14140#issuecomment-1159470071

Libman still use old endpoint:
https://github.com/aspnet/LibraryManager/blob/8dd0ac5fdfe76c2d37ded284834c67633cc18d7f/src/LibraryManager/Providers/Cdnjs/CdnjsCatalog.cs#L23

**Comment by duykhiem** [comment]
> possibly related to cdnjs API change, see [cdnjs/cdnjs#14140](https://github.com/cdnjs/cdnjs/issues/14140); it's being rolled out partially, which would explain the intermittent failure

yes I think so

**Comment by MattIPv4** [comment]
Hey folks, just wanted to note that after a quick read through the code here, it does seem like this is caused by the deprecation of the `assets` property returning every version on cdnjs.

The fix for libman here would be to use the `versions` property to determine the library versions, not `assets`. And then to use the `api.cdnjs.com/libraries/:library/:version` endpoint when it needs to actually get the assets for a specific version.

**Comment by jthomas81** [comment]
Does someone have an example of syntax to do this? This is what I have been doing:
{
  "version": "1.0",
  "defaultProvider": "cdnjs",
  "libraries": [
    {
      "library": "jquery-validate@1.17.0",
      "destination": "wwwroot/lib/jquery-validate/"
    }
]
}

**Comment by the-doctor1994** [comment]
> Does someone have an example of syntax to do this? This is what I have been doing: 

it's not a change for the end user, MattIPv4 is suggesting LibMan update their code to conform to the altered API behavior.... instead of just reverting the api change that is causing this issue

**Comment by MattIPv4** [comment]
Unfortunately, LibMan is not the only consumer of the cdnjs API, and the deprecation that caused this issue is part of larger ongoing work across the API to improve stability (the cdnjs API often runs 5xx errors for assorted consumers, not just LibMan users). We're not really in a position where we can revert this API change as doing so will only increase the failure rate for the API generally.

**Comment by the-doctor1994** [comment]
I understand the change is necessary and inevitable, but time could and should be given to allow for proper deprecation updates by **all** the users of the API who may be negatively affected (or outright broken) by this change. By your own proposal the deprecation was fast-tracked under the incorrect assumption nobody would be affected...

**Comment by MattIPv4** [comment]
I appreciate what you're saying, and this is why we're approaching this as a slow rollout with only 10% of traffic reaching endpoints that implement this deprecation, so that folks are made aware of if they're impacted by the deprecation (turns out it's pretty impossible to determine who consumes what properties of a public API just looking at unauthenticated GET requests), but can also still easily get around the deprecation by just re-requesting (falling into the 90% of traffic that is unaffected at present).

Until we hear more here in terms of what a fix would look like on LibMan's end, we don't intend to further rollout the change, so it will remain at 10% for now. That being said, ultimately we will have to continue with this deprecation as until we complete this rollout a large number of requests to the API from all consumers, not just LibMan users, will continue to fail unexpectedly.

**Comment by jimmylewis** [comment]
Just quickly jumping in here to say: I'm working on the code change for LibMan, but the rollout will require servicing each version of Visual Studio that includes it (2017, 2019, 2022).  There will be some delays to prepare the LibMan builds and VS servicing changes in each product, and then ultimately it will require customers to install the updates as well.

I don't have timelines yet, I'm focusing first on making the code change for VS2022 and our CLI tool, then I will start investigating the servicing requirements for each VS release.

**Comment by duykhiem** [comment]
I think the safest and fastest way to **work around** this issue is to **include the successfully generated static files** by Libman to our project (when our restore client-side libraries command falls into the 90% of traffic that is unaffected at present) and then **disable Libman**. We can re-enable it when the issue is fixed.

**Comment by ennerperez** [comment]
Also, Azure Pipelines are failing too.
![image](https://user-images.githubusercontent.com/1206858/174907713-f71dbde7-94cf-4815-a9cf-de88b24299ec.png)

**Comment by jimmylewis** [comment]
Alright, for folks using the [CLI tool](https://www.nuget.org/packages/Microsoft.Web.LibraryManager.Cli) or [MSBuild task](https://www.nuget.org/packages/Microsoft.Web.LibraryManager.Build) could you please update to the latest version (2.1.174) and see if you have any issues with that?  From a few projects I've tried, it's working fine.  This will be the fastest way to workaround the issue (at least for unblocking CI or local builds) as the VS servicing cycle will take a longer to sort out. 

Timeline on VS updates still TBD...

**Comment by calvaradocl** [comment]
On our azure pipelines we are getting this error:

`/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018: The "Microsoft.Web.LibraryManager.Build.RestoreTask" task failed unexpectedly. [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018: System.TypeInitializationException: The type initializer for 'Microsoft.Web.LibraryManager.Cache.WebRequestHandler' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Microsoft.Web.LibraryManager.Configuration.Settings' threw an exception. ---> System.IO.FileLoadException: Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621) [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Configuration.Settings..ctor() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Configuration.Settings..cctor() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    --- End of inner exception stack trace --- [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Configuration.ProxySettings.get_Default() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Cache.WebRequestHandler..cctor() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    --- End of inner exception stack trace --- [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Build.Contracts.Dependencies.Initialize() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Web.LibraryManager.Build.RestoreTask.Execute() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/src/Frontend/Frontend.csproj]
/root/.nuget/packages/microsoft.web.librarymanager.build/2.1.174/build/Microsoft.Web.LibraryManager.Build.targets(35,9): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/src/Frontend/Frontend.csproj]
    2 Warning(s)
    1 Error(s)
`

**Comment by stebueh** [comment]
updating the MSBuild task fixed the issue for me building locally, thank you @jimmylewis for the timly fix; our azure devops pipeline is using the CLI tool and was running flawlessly all the time, so I guess it is/was still on the 90% of traffic unaffected by the API change.

**Comment by ardalis** [comment]
> I appreciate what you're saying, and this is why we're approaching this as a slow rollout with only 10% of traffic reaching endpoints that implement this deprecation, so that folks are made aware of if they're impacted by the deprecation (turns out it's pretty impossible to determine who consumes what properties of a public API just looking at unauthenticated GET requests), but can also still easily get around the deprecation by just re-requesting (falling into the 90% of traffic that is unaffected at present).

Running into [Hyrum's Law](https://www.hyrumslaw.com/):

_With a sufficient number of users of an API,
it does not matter what you promise in the contract:
all observable behaviors of your system
will be depended on by somebody._

**Comment by farhadmammadli** [comment]
cdnjs issue was resolved in new update.
the updated version is [2.1.174](https://www.nuget.org/packages/Microsoft.Web.LibraryManager.Build/2.1.174)

**Comment by jimmylewis** [comment]
@thiago-bessa try again after running `libman cache clean`?  We use the cache library metadata to disk for 1 day (IIRC of the top of my head) to avoid hitting the service repeatedly on queries, so if you have recently cached metadata from the older build, it might be conflicting with the changes.

I was able to restore fine with this libman.json file
```
{
  "version": "1.0",
  "defaultProvider": "cdnjs",
  "libraries": [
    {
      "library": "jquery@2.2.4",
      "destination": "lib\\jquery"
    },
    {
      "library": "jquery-validate@1.19.1",
      "destination": "lib\\jquery-validate"
    }
  ]
}
```

**Comment by thiago-bessa** [comment]
Sorry @farhadmammadli and @jimmylewis, I didn't copy/paste everything.
![image](https://user-images.githubusercontent.com/9695889/175124611-6fa20306-cdcd-4b5e-a90d-b09da5f5e4f4.png)

I did run `libman cache clean` before running `libman restore`, and it still failed.

Now I ran it an hour later (same cmd prompt, with the same project), and it could restore everything. 

Then I ran it in other project of ours, and it again failed to resolve:
![image](https://user-images.githubusercontent.com/9695889/175125556-42b585a9-30a0-4977-a25d-4464b353c456.png)

It seems kind of intermittent.

**Comment by alex-jitbit** [comment]
I have the same issue, and clearing cache helps. Which probably means the issue is inside libman, not cdnjs?

**Comment by Cloudfare123** [comment]
Our Web solutions all fixed now locally and CI by updating to latest package.  👍 

![image](https://user-images.githubusercontent.com/75725301/175299163-e1a026e0-6ef4-4fbb-af64-f4715f13b8aa.png)

**Comment by easlavov** [comment]
> On our azure pipelines we are getting this error after updating the package to the 2.1.174 version, this persists even if we have Newtonsoft.Json installed.

@calvaradocl, we're having the same issue after upgrading to  `2.1.174`. Did you manage to fix it?

**Comment by Smaiil** [comment]
> Our Web solutions all fixed now locally and CI by updating to latest package. 👍
> 
> ![image](https://user-images.githubusercontent.com/75725301/175299163-e1a026e0-6ef4-4fbb-af64-f4715f13b8aa.png)

thank you!

**Comment by WyrmUK** [comment]
The 2.1.175 release fixes a build (when you build the solution or rebuild all) however in Visual Studio 2022 Professional if you right click on the libman.json file and select Restore Client-Side Libraries it can't find any libraries:

The "jquery@3.6.0" library could not be resolved by the "cdnjs" provider

**Comment by jimmylewis** [comment]
Unfortunately, the process for shipping updates in VS is slower than publishing the NuGet packages.  The same fixes will be included in the next servicing releases for VS2019 and VS2022, as well as the next VS2022 preview.

**Comment by WyrmUK** [comment]
However I have another project that is a NetCoreApp3.1 based web app where even 2.1.175 doesn't fix the issue. It won't list anything other than 6.1.1 for font-awesome and all libraries state something similar to:

The "jquery@3.6.0" library could not be resolved by the "cdnjs" provider

This is basically killing the use of this. I know several large users who have decided to cut their losses and stop using Libman and CD systems.

**Comment by jimmylewis** [comment]
> It won't list anything other than 6.1.1 for font-awesome

@WyrmUK can you provide the steps you're using to list the versions or share your libman.json file?  Installing a newer NuGet package won't impact the experience in VS (Add New Client-side Library wizard or install on save in libman.json), those will need to wait for the VS update.  However, the CLI and build tasks should be agnostic to the cdnjs API change (i.e. should work with responses with or without the change), as we stopped relying on the assets property entirely.

**Comment by WyrmUK** [comment]
It appears that was a regional API thing presumably due to CDNJS.
When on a VPN: editing the libman.json file wouldn't retrieve anything other than 6.1.1 for font-awesome and all libraries were unable to be resolved when building.
When not on the VPN all was fine.
This wasn't a 10% thing as the build was attempted many times.
CDNJS presumably have fixed things in their API now as on the VPN it is all fine today.
Whoever thought it was a good idea to a) remove functionality at a moments notice and b) to implement changes to only some servers needs to be removed from the project - who knows what bad decisions they will make in the future.

**Comment by xtuc** [comment]
@WyrmUK  b) we did not implement changes only on some servers.

**Comment by WyrmUK** [comment]
Rather than spend time issuing denials, perhaps you should spend more time making good decisions in the first place? The confidence in CDNJS has been significantly dented because of all this.

**Comment by MattIPv4** [comment]
As I've explained further up in this thread, this API change was made as part of work to stabilise the API as many consumers (LibMan users, as well as others) had been running into random errors and failures with the API. This change was part of work to solve for that, and so had to be made.

Given the scale of the cdnjs API and it being unauthenticated, there is no easy way for us to know who relies on what properties of the API responses, nor an easy touch-point to communicate changes with consumers, hence we chose the slow rollout approach to allow folks time to notice that they may have been relying on the deprecated property and get things updated as such.

As soon as we found out that LibMan was reliant on this property we stopped any further rollout and left the change at 10% to prompt folks to update as soon as possible. I believe the final LibMan patch should be going out today, so once that it out we will continue the rollout once again to finally stabilise the issues with the cdnjs API :)

**Comment by MattIPv4** [comment]
(Also as Sven noted, this change wasn't regional nor have we changed anything since the initial 10% rollout -- it is a globally distributed Worker that is serving 10% of requests randomly.)

**Comment by WyrmUK** [comment]
However this is about libman and not CDNJS so whilst that change could have been handled so much better - I'll leave things there and report back on libman when the fix has rolled out.

**Comment by MattIPv4** [comment]
Thank you for all the help in getting this resolved @jimmylewis 🧡 

We'll leave things as-is over the weekend, and then I suspect we'll start increasing the rollout percentage for the API changes moving into next week, playing it by ear to ensure these updates are working for folks.

**Comment by PromoFaux** [comment]
LGTM, thanks for the fix! 

![image](https://user-images.githubusercontent.com/1998970/179221321-0d94bb28-f41e-45f4-a9cc-6a5490a19428.png)

**Comment by WyrmUK** [comment]
The update seems to have addressed the remaining issues.

I would advise CDNJS to leave things a little longer than a few days before increasing the rollout to give people time to update both Visual Studio and their various projects. Commercial developments on a release phase often take a couple of weeks to complete during which time they will have frozen the codebase and won't look kindly on anything that prevents them building the final release at the end of the phase.

**Comment by MattIPv4** [comment]
> won't look kindly on anything that prevents them building the final release at the end of the phase

I hear your point, but the API is currently rather unstable as-is with random failures happening, and the 10% rollout remains in place, so really the API is already broken from their perspective. Rolling out to 100% feels like the best bet here so that folks are no longer tracking down an inconsistent, but rather a consistent one that's easier to spot and then update to fix.

**Comment by JosephAmalfitanoSSA** [comment]
any ETA?

**Comment by MattIPv4** [comment]
@JosephAmalfitanoSSA The releases with the fix were made available mid July: https://github.com/aspnet/LibraryManager/issues/685#issuecomment-1185272933

**Comment by JosephAmalfitanoSSA** [comment]
@MattIPv4 I noticed that, but i see the ticket is still marked open. I assumed because it wasn't working again.  I'm still running into. 

I've tried:
-  updating VS to latest stable
-  updating VS to latest preview build (17.4.0 preview 1.0)
-  manually installing: `dotnet tool install -g Microsoft.Web.LibraryManager.Cli` and running `libman cache clean` & `libman install`
-  maybe something else i can't remember

All results are the same X library could not be resolved by the "cdnjs" provider.

**Comment by WyrmUK** [comment]
Update VS to latest stable AND update the Library Manager (Microsoft.Web.LibraryManager.Build) to the latest in the project.

**Comment by JosephAmalfitanoSSA** [comment]
> Update VS to latest stable AND update the Library Manager (Microsoft.Web.LibraryManager.Build) to the latest in the project.

Ahh -- that was the "maybe something else i can't remember"...tried as well. If it's working for the general population, then maybe it's a proxy issue on my end then.

**Comment by jimmylewis** [comment]
If you can capture network traces of the requests sent by VS or by the libman tool, we can try to figure out what is going wrong (either the message is getting caught by a proxy, an error being returned from the API, or unexpected response content).

**Comment by mmcdonald47** [comment]
`The type initializer for 'Microsoft.Web.LibraryManager.Cache.WebRequestHandler' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Microsoft.Web.LibraryManager.Configuration.Settings' threw an exception. ---> System.IO.FileLoadException: Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)`

I'm also running into this `Newtonsoft.Json` issue in my CI pipeline, using `Microsoft.Web.LibraryManager.Build 2.1.175`

**Comment by JosephAmalfitanoSSA** [comment]
`Downloading file https://cdnjs....`

My issue was a proxy issue. It looks like if you have a `HTTP_PROXY` and/or `HTTPS_PROXY `value set as an environment variable, then libman traffic ends up routing traffic through it. 

Steps for me to remedy 
1. Renamed those system/environment variables to `xHTTP_PROXY` and `xHTTPS_PROXY`
2. Installed / restored libraries
3. Renamed those system/environment variables back to `HTTP_PROXY` and `HTTPS_PROXY`
🕺

**Comment by jimmylewis** [comment]
@mmcdonald47 can you capture a Fusion log while running the build?  The NuGet package includes a copy of that assembly, so I'm wondering if that log can help clarify why it's not being located.  (See how to capture Fusion logs at https://docs.microsoft.com/en-us/dotnet/framework/tools/fuslogvw-exe-assembly-binding-log-viewer)

**Comment by mmcdonald47** [comment]
@jimmylewis how can I install this in my Azure CI pipeline? I've tried adding a Command Line Script task, but when running the build it failed because it couldn't recognize `fuslogvw`.

![image](https://user-images.githubusercontent.com/89601258/185403857-d463775f-1cab-4316-9d1b-08f8c6901938.png)

(Apologies for my ignorance. I've not had much experience configuring CI pipelines.)

**Comment by jimmylewis** [comment]
@mmcdonald47 I assume this only occurs on your CI and not locally?  In that case, we can try doing it with via the registry, as in https://stackoverflow.com/questions/255669/how-to-enable-assembly-bind-failure-logging-fusion-in-net

The `EnableLog` key will enable the log, and the text will be included in the thrown exception's message.  This might or might not show up in the message you showed above (since it's wrapped, I'm not sure if it will show the full message).
The `LogFailures` key will enable logging failures to a file, and `LogPath` is (obviously) the log path.  Make sure you remove these keys afterwards, or else it will affect any other assembly loads on the machine (and can eventually consume the remaining disk space).
For testing purposes, you could set these then add the LogPath as a build artifact, and see if there's a logged file for the Newtonsoft.Json binding failure.

**Comment by mmcdonald47** [comment]
@jimmylewis per that SO post I was able to locate the Fusion Log Viewer and get some more information after running my application locally. In checking the logs I see this message

`The operation failed.
Bind result: hr = 0x80004005. Unspecified error

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Program Files\IIS Express\iisexpresstray.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Numerics, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
 (Fully-specified)
LOG: Appbase = file:///C:/Program Files/IIS Express/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = iisexpresstray.exe
Calling assembly : Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed.
===
LOG: Start validating all the dependencies.
LOG: [Level 1]Start validating native image dependency mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089.
Rejecting native image because native image dependency C:\WINDOWS\Microsoft.Net\assembly\GAC_64\mscorlib\v4.0_4.0.0.0__b77a5c561934e089\mscorlib.dll had a different identity than expected
LOG: Start validating all the dependencies.
LOG: [Level 1]Start validating native image dependency mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089.
Native image has correct version information.
LOG: Validation of dependencies succeeded.
LOG: Bind to native image succeeded.`

I'm not entirely sure what this means, but it does mention `Newtonsoft.Json 13.0.0.0`.

**Comment by duykhiem** [comment]
> The changes to Visual Studio are now available in the 16.9.23, 16.11.17, 17.0.12, 17.2.6, and 17.3 Preview 3 releases. Please let me know if you are still having issues after updating to one of these versions. In my testing with each of them, I was able to install the latest and non-latest versions of multiple packages and did not encounter any issues.

Updating to ```Microsoft.Web.LibraryManager.Build.2.1.175``` and ```Visual Studio 2022 version 17.3.1``` fixed this issue for me.
Thanks @jimmylewis :)

**Comment by mentormate-ivandzhunov** [comment]
I also have the issue with Newtonsoft.Json, Version=13.0.0.0 during deployments. Did anyone found a solution for it?

**Comment by WyrmUK** [comment]
If you are using Newtonsoft directly then upgrade it to the latest v13. If you are including another package that uses Netwonsoft then upgrade that. Or add Newtonsoft v13 and set it to be used for all versions. Newtonsoft less than v13 has a vulnerability so should no longer be used. This is not on Topic though and there is enough help for mismatched package versions on stack overflow etc.

**Comment by mmcdonald47** [comment]
@mentormate-ivandzhunov you can try the solution in this StackOverflow post: https://stackoverflow.com/questions/46111749/adding-a-bindingredirect-to-a-net-standard-library/46120907#46120907

My solution was to uninstall LibraryManager, though that doesn't seem like a real solution to me.

**Comment by paillave** [comment]
At my side, in my csproj, I made the following change, and it finally works gracefully!!!

from
```
<PackageReference Include="Microsoft.Web.LibraryManager.Build" Version="2.1.161" />
```

to
```
<PackageReference Include="Microsoft.Web.LibraryManager.Build" Version="2.1.175" />
```

**Comment by paillave** [comment]
> > The changes to Visual Studio are now available in the 16.9.23, 16.11.17, 17.0.12, 17.2.6, and 17.3 Preview 3 releases. Please let me know if you are still having issues after updating to one of these versions. In my testing with each of them, I was able to install the latest and non-latest versions of multiple packages and did not encounter any issues.
> 
> Updating to `Microsoft.Web.LibraryManager.Build.2.1.175` and `Visual Studio 2022 version 17.3.1` fixed this issue for me. Thanks @jimmylewis :)

I confirm

**Comment by duykhiem** [comment]
This issue is fixed. 
I'm going to close it now. 

Please create another bugs for other issue!

Thank you very much! 😄

**Comment by GeorgeMaharis123** [comment]
Is there a workaround for VS2017? The issue is still present there with Microsoft.Web.LibraryManager.Build.2.1.175

**Comment by duykhiem** [comment]
@GeorgeMaharis123  Did you update VS2017 to the latest version?

**Comment by jimmylewis** [comment]
@GeorgeMaharis123 unfortunately we hadn't maintained the build infrastructure for VS2017 over the last several years, and our CI configurations aren't currently able to build the `rel/vs2017` branch without making some significant changes.  However, as a workaround, you should be able to build and install from the tip of that branch with the fix.  A locally built VSIX should be able to install on top of the one that comes as part of VS.

This is the first request we've received about this issue on VS2017; our telemetry showed very little remaining usage in that version of VS, so we originally postponed the effort to get the build pipeline running again.  I'll see if I can request some resources (i.e. time) to get the CI build running again to put the fix into a servicing patch for VS2017.

**Comment by GeorgeMaharis123** [comment]
@duykhiem Yes, i am using version 15.9.50

**Comment by LLIAMAH** [comment]
Visual Studio 2022 Latest updates
![image](https://user-images.githubusercontent.com/4736499/195197879-75294fa8-ceef-4648-bd74-083280b15a9a.png)

Problem not fixed. Cache cleaning does not help. Any ideas? Help?
![image](https://user-images.githubusercontent.com/4736499/195198072-38172a90-ee48-4c0f-8ac6-376f59340965.png)

**Comment by jimmylewis** [comment]
@LLIAMAH can you get a network trace of the requests and responses?

**Comment by LLIAMAH** [comment]
> @LLIAMAH can you get a network trace of the requests and responses?

I took fiddler stack - if you need some special info please guide me which props i've to switch on to help you to investigate problem.

https://github.com/LLIAMAH/CDNPoblemVS2022/blob/master/FiddlerStack.txt

**Comment by jtbrower** [comment]
Hello Everyone, I believe that my companies stable Azure WebApp is now hitting this issue based on what I am seeing.  I will try to provide as much information as possible, but I was actually working hot on another project and accidentally discovered a dead link on our company website and thought this would be a quick fix, but I have been on this for hours without luck. 

**The Error**
_Note : I replaced the source path with the word redacted to hide company details._

`libman.json : error LIB002: The "jquery@3.6.0" library could not be resolved by the "cdnjs" provider [redacted.csproj]`
`libman.json : error LIB002: The "choices.js@9.0.1" library could not be resolved by the "cdnjs" provider [redacted.App.csproj]`

** Most of What I have Tried and I got the Same Error**

- Ran dotnet tool install -g Microsoft.Web.LibraryManager.Cli, then libman cache clean & libman restore
- Tried building through powershell using cli then switched back to troubleshooting with Visual Studio
- I was using an older vs2022 preview build so I updated it to version 17.4.0 Preview 3
- I was using an older vs2022 stable build, then updated to 17.3.6
- I tried using vs2019 v16.11.11 then updated that to v16.11.20
- Opened the the csproj file for the app and updated Microsoft.Web.LibraryManager.Build from 2.1.113 to 2.1.175
- Intellicode suggested that I reorder my properties so I let it place the defaultDestination ahead of defaultProvider 
- Deleted bin and obj directories and a handful of other voodoo "standing on one foot and patting my head" nonsense

**Beyond That**
... I can't promise I will be able to help  debug this much further.  As I noted I should be working on something else and didn't expect that changing a deadlink on a stable WebApp would be such a major problem.  If this is truly closed and fixed and I am doing something odd, forgive me.  Otherwise, I don't see how this issue is resolved?
